docs: ABCI 2.0 tutorial (#1374)
cometbft/cometbft: docs/tutorials/forum-application/1.abci-intro.md

# Introduction to ABCI 2.0

`ABCI 2.0` is an updated version of the ABCI (Application Blockchain Interface) from CometBFT. For more details, check the [specification document](https://docs.cometbft.com/v1.0/spec/abci/).

![ABCI 2.0](images/ABCI2.jpg)

## ABCI 2.0 Methods

### PrepareProposal

- **Purpose**: Modify block proposals before finalization.

> - **Process**: The block proposer collects pending transactions, creating a "raw proposal." The application can then reorder, add, or remove transactions, creating a "prepared proposal."

Reviewer: ```suggestion
- **Process**: The block proposer collects pending transactions, creating a "raw proposal". The application can then reorder, add, or remove transactions, creating a "prepared proposal."
```

### ProcessProposal

- **Purpose**: Validate a block proposal based on application-specific rules.

> - **Process**: Validators assess the proposal and can reject invalid blocks. If rejected, the consensus algorithm will perform a nil prevote. The logic must be deterministic to ensure consistent outcomes.

Reviewer: ```suggestion
- **Process**: Validators assess the proposal and can reject invalid blocks. If rejected, the consensus algorithm will perform a nil prevote. The logic must be deterministic to avoid liveness problems.
```

### ExtendVote

- **Purpose**: Add application-specific data to pre-commit messages.
- **Process**: Validators can add vote extensions to their pre-commit messages, which the application can use later. If no data is added, a zero-length byte array is returned. This method can use non-deterministic logic.

### VerifyVoteExtension

- **Purpose**: Validate vote extension data in pre-commit messages.

> - **Process**: Checks the integrity of vote extensions. If invalid, the entire pre-commit message is rejected. This method must follow deterministic logic.

Reviewer: ```suggestion
- **Process**: Checks the integrity of vote extensions. If invalid, the entire pre-commit message is rejected. This method must follow deterministic logic to avoid liveness problems.
```

### FinalizeBlock

- **Purpose**: Process a decided proposal and update the application's state.

> - **Process**: Includes transaction results, updates to the validator set, changes to consensus parameters, and the `app_hash`. Changes are not persisted until the `Commit` phase.

Reviewer: The fields you are describing here are `FinalizeBlockResponse`. But there's also `FinalizeBlockRequest`, which contains, among others, the transactions to execute, and evidence of misbehaviour. You may also want to mention those.

> ### Commit

Reviewer: Since you're explaining `Commit`, should you also explain `CheckTx`?

- **Purpose**: Permanently store finalized state changes.

> - **Process**: Before `Commit`, the mempool is locked to prevent new transactions. The application must persist state changes to ensure blockchain integrity and reliability.

Reviewer: ```suggestion
- **Process**: Before `Commit`, the mempool is locked to prevent new transactions from reaching the application. The application must persist state changes to ensure blockchain integrity and reliability.
```

reorder transactions, add new ones, or remove unsuitable ones in this draft to optimize the block, resulting in a "prepared proposal".

### ProcessProposal

The `ProcessProposal` method permits a validator to assess a block proposal using specific criteria determined by the
application. This functionality allows for the immediate execution of blocks and allows the application to reject any
invalid blocks, which is crucial for maintaining network integrity. Suppose a proposal is rejected due to invalid content.
In that case, the consensus algorithm will then proceed with a nil prevote, a decision that has significant implications
for the liveness of the CometBFT system.

Although the application is generally expected to accept a prepared proposal, it can selectively ignore invalid elements,
like erroneous transactions, at the time of block execution. The processing logic in `ProcessProposal` must be deterministic,
ensuring that outcomes are consistent and predictable.

### ExtendVote

The `ExtendVote` method allows validators to add non-deterministic, application-specific

> data—called vote extensions—to pre-commit messages, the final voting stage. Not interpreted by the consensus algorithm,

Reviewer: ```suggestion
data —called vote extensions— to pre-commit messages, the final voting stage. Not interpreted by the consensus algorithm,
```

> these extensions are sent alongside the votes and become accessible to the application in future rounds when that validator

Reviewer: ```suggestion
these extensions are sent alongside the votes and become accessible to the application in the next height when that validator
```

is the proposer.

`ExtendVote` is invoked when a non-nil pre-commit message is about to be sent. The application returns a
zero-length byte array if no extension data is available. This feature provides flexibility, as the logic in ExtendVote
can be non-deterministic, tailoring how data is added based on application needs.

### VerifyVoteExtension

The `VerifyVoteExtension` method checks the integrity of data in precommit messages,

> known as vote extensions, which are from a previous block height. If a vote extension fails validation, the entire

Reviewer: ```suggestion
known as vote extensions, which are attached to precommit message received from other validators for the current height and round.
If a vote extension fails validation, the entire
```

precommit message is deemed invalid, hindering the system's ability to finalize blocks.

It's critical to implement this method carefully to prevent disruptions in block finalization. Typically, an application

> should accept an invalid vote extension to maintain consensus but ignore it during operational processing. This verification

Reviewer: ```suggestion
should accept an invalid vote extension to ensure consensus liveness but ignore it during operational processing. This verification
```

is triggered when a precommit message with a vote extension is received for the current height, and must follow deterministic
logic to ensure reliable outcomes.

### FinalizeBlock

The `FinalizeBlock` method processes a decided proposal and provides a comprehensive
response. It includes the results of each transaction, associated events in their original sequence, updates to the
validator set (planned for two heights ahead), and any necessary changes to consensus parameters. The response also
contains the `app_hash`, which verifies the deterministic execution of transactions using an algorithm the application
selects, which ensures the integrity and consistency of the blockchain's state transitions.

It's important to note that while `FinalizeBlock` updates the application's state, these changes do not persist;
this persistence occurs during the `Commit` phase, ensuring proper synchronization and storage of the blockchain data.

### Commit

The `Commit` method is critical for securing the finalized state changes by permanently

> storing them in durable storage. Before `Commit`, CometBFT locks the mempool, ensuring no new transactions are processed

Reviewer 2: good feedback, I've simplified it.

## ABCI 2.0 Methods

### InitChain

- **Purpose**: Initialize the blockchain state.

> - **Process**: Called when a blockchain node starts up for the first time. It sets up the initial state, including the genesis block, consensus parameters and initial validator set.

Reviewer: ```suggestion
- **Process**: Called when a blockchain node starts up for the first time. It sets up the initial state, including the consensus parameters and initial validator set.
```

### Info

- **Purpose**: Provide information about the application.
- **Process**: Called to retrieve information such as the application version, the last block height, and the application state hash.

### Query

- **Purpose**: Query the application state.
- **Process**: Allows external clients to query the application for specific data, such as transaction details.

### CheckTx

- **Purpose**: Validate an individual transaction before it is added to the mempool.

> - **Process**: Every node that receives a transaction runs this method to ensure a transaction is valid.

Reviewer: ```suggestion
- **Process**: Every node that receives a transaction for the first time runs this method to ensure a transaction is valid.
```
